### PR TITLE
Add Key Vault service into fog-azure-rm

### PR DESCRIPTION
**Pull request by vishrutshah**
As part of understanding the `fog-azure-rm` source code and model, I was building a new Key Vault service into fog-azure-rm. 

It's be great to get you guys feedback & suggestions on this PR. We'd appreciate it.

Thanks,

### CODE REVIEW

File: lib/fog/azurerm/models/key_vault/vault.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        attribute :tenant_id
+        attribute :sku_family
+        attribute :sku_name
+        attribute :access_policies
+        attribute :enabled_for_deployment
+        attribute :enabled_for_disk_encryption
+        attribute :enabled_for_template_deployment
+
+        def self.parse(vault)
+          puts vault
```
Comment (bilal-naeem-confiz) [inline lib/fog/azurerm/models/key_vault/vault.rb]: Remove this puts statement

File: lib/fog/azurerm/models/key_vault/vault.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            vault_hash['vault_uri'] = vault_properties.vault_uri
+            vault_hash['tenant_id'] = vault_properties.tenant_id
+
+            unless vault_properties.sku.nil?
+              vault_hash['sku_family'] = vault_properties.sku.family
+              vault_hash['sku_name'] = vault_properties.sku.name
+            end
+
+            vault_hash['access_policies'] = []
+            vault_properties.access_policies.each do |access_policy|
```
Comment (bilal-naeem-confiz) [inline lib/fog/azurerm/models/key_vault/vault.rb]: Check if access policies are nil and then proceed to access the values

File: lib/fog/azurerm/models/key_vault/vault.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+          merge_attributes(Vault.parse(vault))
+        end
+
+        def destroy
+          service.delete_vault(resource_group, name)
+        end
+
+        private
+
+        def validate_input(attr_hash)
```
Comment (bilal-naeem-confiz) [inline lib/fog/azurerm/models/key_vault/vault.rb]: If this method is not being used and has no body then it is best to remove it

File: lib/fog/azurerm/requests/key_vault/create_or_update_vault.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                     'secrets' => ['all']
+                   }
+                }
+              ],
+              'enabledForDeployment' => false,
+              'vaultUri' => 'https =>//samplevault758347.vault.azure.net/'
+            }
+          }
+          vault_mapper = Azure::ARM::KeyVault::Models::Vault.mapper
+          @kay_vault_client.deserialize(vault_mapper, vault, 'result.body')
```
Comment (bilal-naeem-confiz) [inline lib/fog/azurerm/requests/key_vault/create_or_update_vault.rb]: Typo in ``` @kay_vault_client ```
Should be ``` @key_vault_client ```

File: lib/fog/azurerm/requests/key_vault/get_vault.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                    'secrets' => ['all']
+                  }
+                }
+              ],
+              'enabledForDeployment' => false,
+              'vaultUri' => 'https =>//samplevault758347.vault.azure.net/'
+            }
+          }
+          vault_mapper = Azure::ARM::KeyVault::Models::Vault.mapper
+          @kay_vault_client.deserialize(vault_mapper, vault, 'result.body')
```
Comment (bilal-naeem-confiz) [inline lib/fog/azurerm/requests/key_vault/get_vault.rb]: Typo in ``` @kay_vault_client ```
Should be ``` @key_vault_client ```

File: lib/fog/azurerm/requests/key_vault/list_vaults.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                    }
+                  }
+                ],
+                'enabledForDeployment' => false,
+                'vaultUri' => 'https =>//samplevault758347.vault.azure.net/'
+              }
+            }
+          ]
+          vaults_mapper = Azure::ARM::KeyVault::Models::VaultListResult.mapper
+          @kay_vault_client.deserialize(vaults_mapper, vaults, 'result.body')
```
Comment (bilal-naeem-confiz) [inline lib/fog/azurerm/requests/key_vault/list_vaults.rb]: Typo in ``` @kay_vault_client ```
Should be ``` @key_vault_client ```

**Comment by bilal-naeem-confiz** [comment]
Hi @vishrutshah 

These are the changes in the code-base. We suggest you to fix these in the relevant places

1. There are no unit tests. Please write unit tests for the code and ensure maximum code coverage.

2. The complete namespace of a model should be written instead of the model name. This ensures that there will be no conflicts with existing providers written in fog e.g. instead of ``` Vault ``` write ``` Fog::KeyVault:::AzureRM:::Vault ``` in [this](https://github.com/fog/fog-azure-rm/pull/217/files#diff-a4186909b6a31db99602bf50a7e07148) file

3. Integration tests contain all the functionality that is provided to the user. Make sure to write use cases of each functionality in the integration tests. This includes writing use case for ``` get ```, ``` list ``` and ``` delete ``` requests for KeyVault.

4. Add proper documentation for the feature that you have added to fog. This documentation should be as a markdown file and should be added to ``` lib/fog/azurerm/docs ```

5. In most models implemented in fog, we use ` name ` as the identity of the model. Please use ` name ` as the identity in [models/key_vault/access_policy_entry.rb](https://github.com/fog/fog-azure-rm/pull/217/files#diff-51460fa55f5fb6f0f461bb502ab6acc8)

Other issues are specified as comments on the code. We will review your changes as soon as you are done.

Thanks

**Comment by vishrutshah** [comment]
@bilal-naeem-confiz Awesome .Thanks you very much for taking time to go through the first pass. I'll work on addressing the comments and update this PR soon. Thanks a lot again!

**Comment by vishrutshah** [comment]
Hello @bilal-naeem-confiz,

I am still working on writing more unit tests and increasing the coverage & I've addressed some of the issues you mentioned in review and points. I have couple of questions on the 5 points that you mentioned:

Regarding point 5, [AccessPolicyEntry](https://github.com/Azure/azure-sdk-for-ruby/blob/master/management/azure_mgmt_key_vault/lib/generated/azure_mgmt_key_vault/models/access_policy_entry.rb) model of `Azure::ARM::KeyVault::Models` dows not have any attribute named as `name`, could you please advise how can i then use `name` as the identity of the model if there is a way or what should we prefer in that case?

Regarding point 2, could you please elaborate as I am not clear on how should i use full namespace & in which file? 

Thanks!

**Comment by bilal-naeem-confiz** [comment]
Hi @vishrutshah 

Regarding point 2, what I mean is that instead of just using the model name like done [here](https://github.com/fog/fog-azure-rm/pull/217/files#diff-a4186909b6a31db99602bf50a7e07148R35) you should instead use it like ``` Fog::KeyVault::AzureRM::Vault.new ``` The reason for using this is to avoid conflicts with other cloud providers integrated with fog.
Regarding point 5, even if the Azure Ruby SDK does not show ``` name ``` as an attribute, the attribute exists for all resources. Attributes like ``` name ```, ``` resource_group ``` and ``` location ``` are all part of ``` MsRestAzure::Resource ``` and are inherited by all models present in the SDK.

I hope this answers any queries you had. Feel free to ask about any other issues that arise.

Thanks

**Comment by vishrutshah** [comment]
@bilal-naeem-confiz Thank you so much for the quick response. 

I've added unit tests for models & requests, added usage documentations and updated integration tests. I've also fixed the comments and feedbacks you have provided on the review. 

Now, that only think is I am still having trouble understating point 5. :) I understand that you are mentioning that attributes like `name`, `resource_group` and `location` are all part of [MsRestAzure::Resource ](https://github.com/Azure/azure-sdk-for-ruby/blob/master/runtime/ms_rest_azure/lib/ms_rest_azure/resource.rb#L9). The way I am understanding is each model's properties are supposed to go on-wire to the server but that doesn't seem to be the case here because we are not able create [AccessPolicyEntry ](https://github.com/Azure/azure-sdk-for-ruby/blob/master/management/azure_mgmt_key_vault/lib/generated/azure_mgmt_key_vault/models/access_policy_entry.rb#L12) with name. So if we are not going to send it on the wire & not required by the server for AccessPolicyEntry, why should we require `name` from the user / having it as identity to build upon? May be I am not getting the point of having identity in fog but it'd be great if you can please help me with examples to understand. :)

Also, Feel free to review & provide more feedbacks as you guys get a chance. Thanks!

**Comment by bilal-naeem-confiz** [comment]
Hi @vishrutshah 

I see that you have only written unit test for ``` create_or_update ``` request. Please write unit tests for other requests such as ``` get ```, ``` list ``` and ``` delete ``` to ensure maximum code coverage.

Thanks

**Comment by vishrutshah** [comment]
@bilal-naeem-confizthanks for catching that. I totally missed that. Added those tests. Thanks!

**Comment by vishrutshah** [comment]
Hi @bilal-naeem-confiz, I've rebased my changes on top of master due to conflicts. Please review and let me know your views on this. 

Thanks,
Vishrut Shah

**Pull request merged**